icu4c/source/i18n/unitsdata.cpp

```diff
@@ -0,0 +1,194 @@
+        int32_t lenSource = uprv_strlen(source);
+        const UChar *baseUnit = NULL, *factor = NULL, *offset = NULL;
+        int32_t lenBaseUnit, lenFactor, lenOffset;
+        const char *key;
+        for (int32_t i = 0; conversionRateTable.getKeyAndValue(i, key, value); ++i) {
+            if (uprv_strcmp(key, "target") == 0) {
+                baseUnit = value.getString(lenBaseUnit, status);
+            } else if (uprv_strcmp(key, "factor") == 0) {
+                factor = value.getString(lenFactor, status);
+            } else if (uprv_strcmp(key, "offset") == 0) {
+                offset = value.getString(lenOffset, status);
+            }
+        }
+        if (U_FAILURE(status)) return;
+        if (baseUnit == NULL || factor == NULL) {
+            // We could not find a usable conversion rate.
+            status = U_MISSING_RESOURCE_ERROR;
+            return;
+        }
+
+        // Check if we already have the conversion rate in question.
+        //
+        // TODO(review): We could do this skip-check *before* we fetch
+        // baseUnit/factor/offset based only on the source unit, but only if
+        // we're certain we'll never get two different baseUnits for a given
+        // source. This should be the case, since convertUnit entries in CLDR's
+        // units.xml should all point at a defined base unit for the unit
+        // category. I should make this code more efficient after
+        // double-checking we're fine with relying on such a detail from the
+        // CLDR spec?
+        fLastBaseUnit.clear();
+        fLastBaseUnit.appendInvariantChars(baseUnit, lenBaseUnit, status);
+        if (U_FAILURE(status)) return;
+        for (int32_t i = 0, len = outVector.length(); i < len; i++) {
+            if (strcmp(outVector[i]->sourceUnit.data(), source) == 0 &&
+                strcmp(outVector[i]->baseUnit.data(), fLastBaseUnit.data()) == 0) {
+                return;
+            }
+        }
+
+        // We don't have this ConversionRateInfo yet: add it.
+        ConversionRateInfo *cr = outVector.emplaceBack();
+        if (!cr) {
+            status = U_MEMORY_ALLOCATION_ERROR;
+            return;
+        } else {
+            cr->sourceUnit.append(source, lenSource, status);
+            cr->baseUnit.append(fLastBaseUnit.data(), fLastBaseUnit.length(), status);
+            cr->factor.appendInvariantChars(factor, lenFactor, status);
+            if (offset != NULL) cr->offset.appendInvariantChars(offset, lenOffset, status);
+        }
+    }
+
+    /**
+     * Returns the MeasureUnit that was the conversion base unit of the most
+     * recent call to put() - typically meaning the most recent call to
+     * ures_getAllItemsWithFallback().
+     */
+    MeasureUnit getLastBaseUnit(UErrorCode &status) {
+        return MeasureUnit::forIdentifier(fLastBaseUnit.data(), status);
+    }
+
+  private:
+    MaybeStackVector<ConversionRateInfo> &outVector;
```

Comment:
Storing references as fields is a bit smelly IMHO.  C++ doesn't give you a way to detect dangling pointers when you only store a reference.

Reply:
This is related to Hugo pull request: #31

Reply:
Currently the sink is populating a MaybeStackVector it doesn't own. I want to see what kind of ownership transfer options there are.

Prior art review:

- ListFormatter::ListPatternsSink has a public char* read directly by its clients.
- RelDateFmtDataSink stores a pointer (`URelativeString *fDatesPtr;`),
- RelDateTimeFmtDataSink holds a reference passed to the constructor (`RelativeDateTimeCacheData &outputData;`).
- DecFmtSymDataSink holds a reference (`DecimalFormatSymbols& dfs;`)
- PluralTableSink holds a pointer (`UnicodeString *outArray;`)
- NumberRangeDataSink holds a reference (`NumberRangeData& fData;`), as does PluralRangesDataSink.
- ZNames::ZNamesLoader owns the loaded data and expects the client code to make a copy as needed.
- KeywordsSink (ucol_res.cpp) holds a pointer (`UList *values;`)

I'm not quite convinced either way - though stylistically I prefer a pointer, for something that's more clearly "possibly bad" than a reference.

Reply:
I've switched to a pointer and documented the ownership semantics a little (in the construtor).

Not marking as resolved, for the bullet-points survey above, and awaiting Shane's further input: I didn't find neat examples of transfer of ownership, but maybe there are good ways.

Reply:
Moved this here: https://github.com/sffc/icu/pull/31/files#r408353252
Marking as resolved.